File: knn/julia/knn.jl
# Structure to represent a data point
struct DataPoint
    x::Float64
    y::Float64
    label::Int
end

# Function to calculate Euclidean distance between two data points
function calculate_distance(p1::DataPoint, p2::DataPoint)
    dx = p2.x - p1.x
    dy = p2.y - p1.y
    return sqrt(dx * dx + dy * dy)
end

# Function to perform KNN classification
function classify(training_data::Vector{DataPoint}, test_point::DataPoint, k::Int)
    # Calculate distances to all training data points
    distances = [calculate_distance(data_point, test_point) for data_point in training_data]

    # Sort the distances in ascending order
    sorted_distances = sort(distances)

    # Count the occurrences of each label among the k nearest neighbors
    label_count = Dict{Int, Int}()
    for i in 1:k
        index = findfirst(==(sorted_distances[i]), distances)
        label = training_data[index].label
        label_count[label] = get(label_count, label, 0) + 1
    end

    # Return the label with the highest count
    max_count = maximum(values(label_count))
    predicted_label = findfirst(count -> count == max_count, values(label_count))
    return predicted_label
end

# Training data
training_data = [
    DataPoint(2.0, 4.0, 0),
    DataPoint(4.0, 6.0, 0),
    DataPoint(4.0, 8.0, 1),
    DataPoint(6.0, 4.0, 1),
    DataPoint(6.0, 6.0, 1)
]

# Test data point
test_point = DataPoint(5.0, 5.0, 0)

# Perform KNN classification
k = 3
predicted_label = classify(training_data, test_point, k)

# Print the predicted label
println("Predicted label: $predicted_label")

File: knn/matlab/knn.m
% Structure to represent a data point
DataPoint = struct('x', 0, 'y', 0, 'label', 0);

% Function to calculate Euclidean distance between two data points
function distance = calculateDistance(p1, p2)
    dx = p2.x - p1.x;
    dy = p2.y - p1.y;
    distance = sqrt(dx * dx + dy * dy);
end

% Function to perform KNN classification
function predictedLabel = classify(trainingData, testPoint, k)
    % Calculate distances to all training data points
    distances = zeros(length(trainingData), 1);
    for i = 1:length(trainingData)
        distances(i) = calculateDistance(trainingData(i), testPoint);
    end

    % Sort the distances in ascending order
    sortedDistances = sort(distances);

    % Count the occurrences of each label among the k nearest neighbors
    labelCount = containers.Map('KeyType', 'double', 'ValueType', 'double');
    for i = 1:k
        index = find(distances == sortedDistances(i), 1);
        label = trainingData(index).label;
        if isKey(labelCount, label)
            labelCount(label) = labelCount(label) + 1;
        else
            labelCount(label) = 1;
        end
    end

    % Return the label with the highest count
    [~, maxCountIndex] = max(cell2mat(values(labelCount)));
    predictedLabel = cell2mat(keys(labelCount(maxCountIndex)));
end

% Training data
trainingData = [
    struct('x', 2.0, 'y', 4.0, 'label', 0),
    struct('x', 4.0, 'y', 6.0, 'label', 0),
    struct('x', 4.0, 'y', 8.0, 'label', 1),
    struct('x', 6.0, 'y', 4.0, 'label', 1),
    struct('x', 6.0, 'y', 6.0, 'label', 1)
];

% Test data point
testPoint = struct('x', 5.0, 'y', 5.0, 'label', 0);

% Perform KNN classification
k = 3;
predictedLabel = classify(trainingData, testPoint, k);

% Print the predicted label
fprintf('Predicted label: %d\n', predictedLabel);